enhancements/substitutes-for.md

```diff
@@ -0,0 +1,386 @@
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Open Questions [optional]
+
+1. Can we get the operator pipeline using a single opm binary that is always the latest
+version (for both DB building and including in catalogs)?  If not, why not? (it means we 
+must backport this feature to all opm versions the operator pipeline uses today)
+
+## Summary
+
+It is currently very difficult to inject a new version of a bundle into the middle of an upgrade graph.  This
+is because existing nodes in the graph may contain explicit "replaces" or "skips" references to other nodes.
+If a new bundle is added to the middle of the graph, replaces/skips references must be updated to include
+the new node.
+
+There is a need to rebuild old versions of bundles to pick up CVE fixes.  Doing so requires that the
+upgrade graph is preserved such that users who update to the CVE-patched bundle can still continue to
+upgrade to newer bundles, just as if they were still on the un-patched bundle.
+
+For example, suppose the original upgrade graph, as specified via "replaces" references, was:
+
+1.0.0->1.0.1
+
+If the patched version is 1.0.1-patched (1.0.1-patched is the 1.0.0 bundle, rebuilt with CVE fixes.  It is not 
+the v1.0.1 version of the operator.  Due to semver semantics this is ordered before 1.0.1, which is what we want), 
+then ideally we want the new graph to be:
+
+[1.0.0 or 1.0.1-patched]->1.0.1
+
+However achieving such a graph would require updating the 1.0.1 bundle to add a "replaces 1.0.1-patched" reference.
+In turn, 1.0.1 can't be updated without changing its version, which means we end up needing something like:
+
+[1.0.0 or 1.0.1-patched or 1.0.1]->1.0.2
+
+(Where the only difference between 1.0.1 and 1.0.2 is the replaces references).
+
+
+To avoid having to update and publish new versions of subsequent bundles in the upgrade graph, we need a way to
+effectively stitch 1.0.1-patched into the graph without changing 1.0.1 (or any other existing bundles that need
+to be able to upgrade from 1.0.1-patched).
+
+This enhancement proposes we do so by allowing a bundle to specify which other bundles it "substitutes for".  This
+will effectively act as a search/replace in the existing bundles for any references to the bundles this bundle 
+substitutes for.  
+
+Continuing the example above, if 1.0.1-patch is annotated as substituting for 1.0.0, then the effective new upgrade 
+graph would be:
+
+1.0.0->1.0.1-patched->1.0.1
+
+
+The implications of this for opm add + the catalog db are that the substitutes-for metadata must be stored in the DB associated w/ each bundle.
+
+The expected flow is:
+
+1) opm add inserts the new bundle following all existing logic, but also adding the substitutes-for metadata
+2) Upon a list-bundles request, the operator-registry should return a substitution-corrected list of bundles based on substitutes replaces/skips references.
```

Comment:
my expectation is that the graph modifications would take place at add time, so no changes to lisbundles would be needed.

Reply:
yeah i muddied this a bit throughout the EP.  It should be possible to do all the mutation add ADD time.  Just keep in mind that you not only have to mutate the bundle you're adding, you also need to apply any substitutions the new bundle is introducing, to content that is already in the DB.  (And due to the transitive nature, that really means reapply the total set of substitutions, both the existing ones and any that are being added w/ the new bundle)

Doing it during listbundles is somewhat simpler from that perspective, but of course at the cost of doing it every single time.

if we agree the "update the DB" approach is preferred, i will make the EP more opinionated towards the "do the mutation during add" instead of potentially doing it during listbundles.

Reply:
+1

Reply:
transitive substitutions are already taken care of at runtime by the resolver, I'm not sure we need to handle it in `opm` (as long as we express substitutes-for in terms of skips/replaces - so maybe I'm begging the implementation a bit). I think that covers both the arch concern (moving decision making to the resolver) and the practical concern (that we want this to work on 4.6 without backports)

Reply:
I think it still needs to be handled in opm.

Suppose we have the rule:
A'' substitutes A'

and we have a bundle B which replaces A.

obviously bundle B is not touched by the substitution rules we have.

Now we add the rule A' substitutes A.

if we just apply the direct rule, then B replaces A'.

What is going to get B updated to `replaces A''` other than an explicit application of the transitive substitution?  I don't see how the resolver is going to achieve that result.  (At best it might achieve the upgrade of A->A'->B, but it won't achieve A''->B)